scripts/filename_to_metadata.py

```diff
@@ -0,0 +1,311 @@
+        os = os.replace("windows", "Windows")
+
+        # normalize os version
+        if os == "Windows":
+            os_version = os_version.replace("XP", "5.1.0")
+            os_version = os_version.replace("7", "6.1.0")
+            os_version = os_version.replace("8.1", "6.3.0")
+            os_version = os_version.replace("8", "6.2.0")
+            os_version = os_version.replace("10", "10.0.0")
+        elif os == "MacOSX":
+            os_version = os_version.replace("El_Capitan", "10.11.0")
+            os_version = os_version.replace("Yosemite", "10.10.0")
+            os_version = os_version.replace("Mavericks", "10.9.0")
+            os_version = os_version.replace("Mountain_Lion", "10.8.0")
+            os_version = os_version.replace("Lion", "10.7.0")
+            os_version = os_version.replace("Snow_Leopard", "10.6.0")
+        if not (re.match("^([0-9]+)\.([0-9]+)\.([0-9]+)$", browser_version)
+            or re.match("^([0-9]+)\.([0-9]+)$", browser_version)
+            or re.match("^([0-9]+)$", browser_version)):
+            os_version = ""
+
+        self.browser = browser
+        self.browser_version = browser_version
+        self.os = os
+        self.os_version = os_version
+        self.platform = platform
+        self.device = device
+
+class RequestFingerprint:
+    def __init__(self):
+        self.clear()
+
+    def clear(self):
+        self.record_tls_version = ""
+        self.tls_version = ""
+        self.ciphersuites = []
+        self.compression_methods = []
+        self.signature_algorithms = []
+        self.extensions = []
+        self.elliptic_curves = []
+        self.ec_point_formats = []
+        self.headers = []
+        self.quirks = []
+        self.parsed = False
+
+    def __str__(self):
+        if not self.parsed:
+            return ""
+        if len(self.compression_methods) > 1:
+            self.quirks.append("compr")
+        return "{version}:{ciphersuites}:{extensions}:{elliptic_curves}:{ec_point_formats}:{headers}:{quirks}".format(
+            version="{:x}".format(int(self.tls_version,16)),
+            ciphersuites=",".join("{:x}".format(int(x,16)) for x in self.ciphersuites),
+            extensions=",".join("{:x}".format(int(x,16)) for x in self.extensions),
+            elliptic_curves=",".join("{:x}".format(int(x,16)) for x in self.elliptic_curves),
+            ec_point_formats=",".join("{:x}".format(int(x,16)) for x in self.ec_point_formats),
+            headers=",".join(self.headers),
+            quirks=",".join(self.quirks))
+
+    def parse(self, filename):
+        self.clear()
+        pdml = subprocess.run(["tshark", "-r", filename, "-T", "pdml"], capture_output=True, encoding='utf-8').stdout
+        # tshark may omit closing tag on incomplete pcaps
+        if '</pdml>' not in pdml:
+            pdml += '</pdml>'
+        root = etree.fromstring(pdml)
+        for pkt in root:
+            for proto in pkt:
+                if proto.get("name") != "ssl":
```

Comment:
Starting with Wireshark 3.0, the protocol "ssl" was renamed to "tls". This affects the field names as well.

As alternative to -Tpdml you can also consider using `-Tjson` which could be parsed with `json.loads` and friends.

Reply:
I'll switch the script to use Wireshark 3.0! I'll also update the script to use `-Tjson`. It requires some special processing b/c `tshark` outputs json maps with duplicate keys (e.g., for `tls.handshake.ciphersuite`), but I should be able to get around that.

Reply:
Yes, if a field has a single value, it is displayed as string, otherwise it will be shown as an array of strings.

If you are just interested in some fields, you can also use the `-e` option to limit the output (values then appear as array, not sure if that is intentional though):
```
tshark -r example.pcapng -T json -e tls.handshake.ciphersuite -e tls.handshake.type
```
This `-e` option also works with other output types (`-Tfields`, etc.). Yet another option potential option is `-J tls`. If used without `-e`, then it will strip the TCP, IP, etc. layers in case you are just interested in TLS.

Reply:
That works very nicely! I just updated the script to use TShark 3.0 with those parsing options.